Add parser for GoCov

pyieh opened the pull request:
<!-- Please describe your pull request here. -->
Adding a parser for GoCov coverage files.

### Testing done
Testing was done in both with unit tests and a test coverage file (based on real coverage files). As well as on a local Jenkins, where a local version of the Jenkins coverage-plugin was installed to use this coverage model. Ran a simple Go pipeline and validated the coverage output was available in the UI. 

Once approved and merged, the corresponding coverage-plugin change can be found here: N/A

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

uhafner commented on src/main/java/edu/hm/hafner/coverage/parser/GoParser.java:
This parser seems to be way to complex for the simple format of the test case? Are you planning to parser different reports? So I am skipping the review on this class right now. It would also make sense to keep the static analysis happy before starting a review...

uhafner reviewed (changes requested):
Thanks for your pull request for this new format!

pyieh commented on src/main/java/edu/hm/hafner/coverage/CoverageParser.java:
```diff
@@ -166,7 +166,7 @@ public final TreeStringBuilder getTreeStringBuilder() {
      *
      * @return the root of the created tree
-     * @throws ParsingException
+     * @throws RuntimeException
```
It's due to `GoParser` extending `CoverageParser` and overriding the `parseReport()` method. `ParsingException` comes from `SecureXmlParserFactory`, but because Go coverage isn't in XML format, it didn't felt misleading to have it throw an XML Parsing exception or a child exception of it. So I created a new exception that was a child at the next level, which was the`RuntimeException`

pyieh commented on src/test/resources/edu/hm/hafner/coverage/parser/go/go-coverage-atomic.out:
```diff
@@ -0,0 +1,12 @@
+mode: atomic
+github.com/example/project/pkg/utils/file1.go:5.12,8.22 3 1
```
The format is:
`<FILE>`:`<start row>`.`<start column>`,`<end row>`.`<end column>` `<# of statements>` `<# of statements covered>`

uhafner replied:
Ok, then this format actually provides only line and instruction (or in your case statement) coverage. That means the counters for the line and branch coverage are either 1/1 or 0/1 (covered/total), see next comment. 

The exact numbers of statements can be used to set the instruction coverage. Since we have only files as elements this coverage needs to be put per file.

pyieh replied:
My mistake, the format is actually:
`<FILE>`:`<start row>`.`<start column>`,`<end row>`.`<end column>` `<# of statements covered>` `<# of statements missed>`
And so that's we have a running tallies of the number of covered & missed statements, then resolve everything at the end.

uhafner replied:
This does not make sense when I look at the test case file? All of your lines actually show a coverage > 0. Or am I missing something?

```
mode: atomic
github.com/example/project/pkg/utils/file1.go:5.12,8.22 3 1
github.com/example/project/pkg/utils/file1.go:10.9,15.23 5 3
github.com/example/project/pkg/utils/file1.go:19.16,20.23 2 0
github.com/example/project/pkg/utils/file1.go:24.3,25.25 1 1
github.com/example/project/pkg/utils/file1.go:30.16,32.25 1 1
github.com/example/project/pkg/db/file2.go:12.7,15.7 1 0
github.com/example/project/pkg/db/file2.go:17.7,22.14 3 2
github.com/example/project/cmd/file3.go:10.12,10.21 1 0
github.com/example/project/cmd/file3.go:15.17,17.27 3 0
github.com/example/project/pkg/test/file4.go:10.12,11.21 1 0
github.com/example/project/pkg/test/file4.go:10.12,11.21 1 1
````

I rewrote the test so it matches my coding style:
https://github.com/jenkinsci/coverage-model/pull/99/commits/3ae1498b697aeb365fa02d80eac0e1f0a49b497b

I will try to refactor the parser as well. What is important and missing: your parser produces INSTRUCTION coverage. Which implies line coverage.

pyieh replied:
Let's take the example line:
```
github.com/example/project/pkg/utils/file1.go:5.12,8.22 3 1
```
It'd be read as:
For `file1.go` , lines `5`-`8`, there were `3` statements that were covered and `1` statement that was missed.

uhafner replied:
But how can then be 
report:
```
github.com/example/project/pkg/db/file2.go:12.7,15.7 1 0
github.com/example/project/pkg/db/file2.go:17.7,22.14 3 2
```
actual test:
````
 two -> assertThat(two).hasName("file2.go")
                        .hasMissedLines(12, 13, 14, 15)
                        .hasCoveredLines(17, 18, 19, 20, 21, 22)
```
From the counters I would expect: Lines 12-15 have one statement covered and 0 missed. So we should have:

````
 two -> assertThat(two).hasName("file2.go")
                        .hasCoveredLines(12, 13, 14, 15, 17, 18, 19, 20, 21, 22)
```

uhafner replied:
Do you need some help to get this parser finished? I think we are currently struggling at the missed/covered computation, see comment above.

pyieh commented on src/test/java/edu/hm/hafner/coverage/parser/GoParserTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    @Test
+    void shouldFailWhenParsingInvalidFiles() {
```
Calling the base class's/super's method will have it check `empty.xml`, where it expects a `NoSuchElementException`, but Go coverage files aren't in XML format, so having it parse that file will result in a different error, thus failing that test.

pyieh commented on src/main/java/edu/hm/hafner/coverage/parser/GoParser.java:
```diff
@@ -0,0 +1,188 @@
+import static edu.hm.hafner.coverage.parser.JacocoParser.createValue;
+
+public class GoParser extends CoverageParser {
+    private static final java.util.logging.Logger LOGGER = java.util.logging.Logger.getLogger(GoParser.class.getName());
```
I'm already are doing so ([example](https://github.com/pyieh/coverage-model/blob/gocov/src/main/java/edu/hm/hafner/coverage/parser/GoParser.java#L37-L38)), I was using these to also log the information to Jenkins logs, but they can be removed to better align with the rest of the module, if you want.

pyieh commented on src/main/java/edu/hm/hafner/coverage/FileNode.java:
```diff
@@ -521,6 +521,17 @@ public FileNode addCounters(final int lineNumber, final int covered, final int m
     }
 
+    @CanIgnoreReturnValue
+    public FileNode addCounterIncremental(final int lineNumber, final int covered, final int missed) {
```
As addressed in previous statement, my mistake on the previous incorrect information. But because each line read in the coverage file gives N number of covered statements, and M number of missed statements, we keep running tallies of each one and then resolve everything at the end.

pyieh commented:
@uhafner Addressed your comments and corrected an incorrect statement I made before. Hopefully the correct explanation of the Go coverage files justifies our methodology for counting coverage.

uhafner commented:
Wouldn't it make sense to use https://github.com/AlekSi/gocov-xml? Then we simply can reuse the Cobertura parser? Wouldn't there be also method names when used in that way?

pyieh commented:
> Wouldn't it make sense to use https://github.com/AlekSi/gocov-xml? Then we simply can reuse the Cobertura parser? Wouldn't there be also method names when used in that way?

@uhafner we tried avoiding using any non-organization type of libraries due to possible lack of future support reasons, but this author seems to be highly involved in the Go community and if you're okay with us using this library, I can re-implement it to do so.

uhafner commented:
> > Wouldn't it make sense to use https://github.com/AlekSi/gocov-xml? Then we simply can reuse the Cobertura parser? Wouldn't there be also method names when used in that way?
> 
> @uhafner we tried avoiding using any non-organization type of libraries due to possible lack of future support reasons, but this author seems to be highly involved in the Go community and if you're okay with us using this library, I can re-implement it to do so.

I'm not sure if I can follow. What I am suggesting is: rather than creating the file `go-coverage-atomic.out` in your build and parsing this file in Jenkins, it looks much easier to call `https://github.com/AlekSi/gocov-xml` in your build pipeline. Then a cobertura.xml file will be created that can be read into Jenkins without any modification in the coverage-model. 

At least it would make sense to use `https://github.com/AlekSi/gocov-xml` in a protoype to see if the created report contains more information than the plain `go-coverage-atomic.out` file? 

So I am not against your parser in this PR but if it is not required because a better tool already exists it would make sense to follow the other approach... 

And if this does not work we need to improve your parser so that the created model is correct.

pyieh commented:
> > > Wouldn't it make sense to use https://github.com/AlekSi/gocov-xml? Then we simply can reuse the Cobertura parser? Wouldn't there be also method names when used in that way?
> > 
> > 
> > @uhafner we tried avoiding using any non-organization type of libraries due to possible lack of future support reasons, but this author seems to be highly involved in the Go community and if you're okay with us using this library, I can re-implement it to do so.
> 
> I'm not sure if I can follow. What I am suggesting is: rather than creating the file `go-coverage-atomic.out` in your build and parsing this file in Jenkins, it looks much easier to call `https://github.com/AlekSi/gocov-xml` in your build pipeline. Then a cobertura.xml file will be created that can be read into Jenkins without any modification in the coverage-model.
> 
> At least it would make sense to use `https://github.com/AlekSi/gocov-xml` in a protoype to see if the created report contains more information than the plain `go-coverage-atomic.out` file?
> 
> So I am not against your parser in this PR but if it is not required because a better tool already exists it would make sense to follow the other approach...
> 
> And if this does not work we need to improve your parser so that the created model is correct.

Unfortunately, we're trying to automate and minimize the number of steps needed for users to migrate over to this new plugin. And users are currently using that `.out` coverage format. 
Can I ask you to clarify what's currently incorrect about my model? I thought with the updated/correct explanation of each line, my model shows the correct coverage data.

uhafner commented:
> > > > Wouldn't it make sense to use https://github.com/AlekSi/gocov-xml? Then we simply can reuse the Cobertura parser? Wouldn't there be also method names when used in that way?
> > > 
> > > 
> > > @uhafner we tried avoiding using any non-organization type of libraries due to possible lack of future support reasons, but this author seems to be highly involved in the Go community and if you're okay with us using this library, I can re-implement it to do so.
> > 
> > 
> > I'm not sure if I can follow. What I am suggesting is: rather than creating the file `go-coverage-atomic.out` in your build and parsing this file in Jenkins, it looks much easier to call `https://github.com/AlekSi/gocov-xml` in your build pipeline. Then a cobertura.xml file will be created that can be read into Jenkins without any modification in the coverage-model.
> > At least it would make sense to use `https://github.com/AlekSi/gocov-xml` in a protoype to see if the created report contains more information than the plain `go-coverage-atomic.out` file?
> > So I am not against your parser in this PR but if it is not required because a better tool already exists it would make sense to follow the other approach...
> > And if this does not work we need to improve your parser so that the created model is correct.
> 
> Unfortunately, we're trying to automate and minimize the number of steps needed for our users to migrate over to this new plugin. And users are currently using that `.out` coverage format. Can I ask you to clarify what's currently incorrect about my model? I thought with the updated/correct explanation of each line, my model shows the correct coverage data.

See my comment https://github.com/jenkinsci/coverage-model/pull/99#discussion_r1598257155

uhafner commented:
Superseded by #169
